src/oemof/solph/_energy_system.py

```diff
@@ -133,6 +161,85 @@ def __init__(
             timeindex=timeindex, timeincrement=timeincrement, **kwargs
         )
 
+        if multi_period:
+            msg = (
+                "CAUTION! You specified 'multi_period=True' for your "
+                "energy system.\n This will lead to creating "
+                "a multi-period optimization modeling which can be "
+                "used e.g. for long-term investment modeling.\n"
+                "Please be aware that the feature is experimental as of "
+                "now. If you find anything suspicious or any bugs, "
+                "please report them."
+            )
+            warnings.warn(msg, debugging.SuspiciousUsageWarning)
+        self.multi_period = multi_period
+        self.periods = self._add_periods(periods, freq)
+        self._extract_periods_years()
+
+    def _add_periods(self, periods, freq="H"):
+        """Returns periods to be added to the energy system
+
+        * For a standard model, periods only contain one value {0: 0}
+        * For a multi-period model, periods are indexed with integer values
+          starting from zero. The keys are the time indices for the resective
+          period given by a pd.date_range object. As a default,
+          each year in the timeindex is mapped to its own period.
+
+        Parameters
+        ----------
+        periods : dict
+            Periods of a (multi-period) model
+            Keys are periods as increasing integer values, starting from 0,
+            values are the periods defined by a pd.date_range object;
+            For a standard model, only one period is used.
+
+        freq : str
+            The frequency of a multi-period model;
+            Defaults to "H"
+
+        Returns
+        -------
+        periods : dict
+            Periods of the energy system (to ensure it being set)
+        """
+        if not self.multi_period:
+            periods = {0: 0}
+        elif periods is None:
+            years = sorted(list(set(getattr(self.timeindex, "year"))))
+
+            periods = {}
+            filter_series = self.timeindex.to_series()
+            for number, year in enumerate(years):
+                start = filter_series.loc[
+                    filter_series.index.year == year
+                ].min()
+                end = filter_series.loc[filter_series.index.year == year].max()
+                periods[number] = pd.date_range(start, end, freq=freq)
+        else:
+            for k in periods.keys():
+                if not isinstance(k, int):
+                    raise ValueError("Period keys must be of type int.")
+
+        return periods
+
+    def _extract_periods_years(self):
```

Comment:
Well yeah, in pricinple it could, but it brings along some follow-up problems:
* handling discounting and annuities which is done within the model
* lifetime tracking where a mapping is made on an annual basis and periods are mapped back to "real" years.